### PR TITLE
Document functions, add features: 'return to menu', 'remove tiles'

### PR DESCRIPTION
**Pull request by BurhanHTW**
- Implemented 'Return to Menu' functionality.
- Added 'Remove Tiles' feature.
- Unified saving game state and statistics into a single file.
- Enabled selection of saved game states for continuation.

### CODE REVIEW

**Comment by BurhanHTW** [comment]
We have submitted the pull request, and it has been a pleasure to contribute to this project.

File: src/.vscode/c_cpp_properties.json [hunk 1]
```diff
@@ -0,0 +1,18 @@
+{
```
Comment (plibither8) [inline src/.vscode/c_cpp_properties.json:1]: Let's remove the `.vscode` files. They're meant for developer-specific configs on local machines, so shouldn't be part of the repository.

Bonus: Add the `.vscode` folder to the `.gitignore` file. This should prevent it from being staged and committed in the future as well! :)

File: src/.vscode/launch.json [hunk 2]
```diff
@@ -0,0 +1,24 @@
+{
```
Comment (plibither8) [inline src/.vscode/launch.json:1]: Same with this file... let's remove it.

File: src/.vscode/settings.json [hunk 3]
```diff
@@ -0,0 +1,114 @@
+{
```
Comment (plibither8) [inline src/.vscode/settings.json:1]: ...and this one too

File: src/game-graphics.cpp [hunk 4]
```diff
@@ -117,11 +117,22 @@ std::string BoardSizeErrorPrompt() {
...
+ * @brief Generates a string prompt listing the available input commands.
+ * 
+ * This function creates a formatted string that lists the available input commands for the game.
+ * The commands include movements (Up, Left, Down, Right), saving the game, and returning to the menu.
+ * The prompt is formatted with indentation for readability.
+ * 
+ * @note The new addition includes 'M' for returning to the menu.
+ * 
+ * @return std::string A formatted string containing the list of input commands.
+ */
```
Comment (plibither8) [inline src/game-graphics.cpp:130]: Comments look great, thank you!

File: src/game-graphics.cpp [hunk 5]
```diff
@@ -117,11 +117,22 @@ std::string BoardSizeErrorPrompt() {
...
 }
 
+/**
+ * @brief Generates a string prompt listing the available input commands.
+ * 
+ * This function creates a formatted string that lists the available input commands for the game.
+ * The commands include movements (Up, Left, Down, Right), saving the game, and returning to the menu.
+ * The prompt is formatted with indentation for readability.
+ * 
+ * @note The new addition includes 'M' for returning to the menu.
```
Comment (plibither8) [inline src/game-graphics.cpp:127]: We can remove this note, since it will get outdated pretty quickly.

File: src/game-pregamemenu.cpp [hunk 6]
```diff
@@ -125,32 +125,40 @@ void SetUpNewGame(NewGameFlag ns) {
...
-  std::tie(loaded_gameboard, tempGBoard) =
-      load_GameBoard_data_from_file(gameboard_data_filename);
-  std::tie(loaded_game_stats, score_and_movecount) =
-      load_game_stats_from_file(game_stats_data_filename);
-  std::tie(tempGBoard.score, tempGBoard.moveCount) = score_and_movecount;
-
-  const auto all_files_loaded_ok = (loaded_gameboard && loaded_game_stats);
-
-  return std::make_tuple(all_files_loaded_ok, tempGBoard);
+ // const auto gameboard_data_filename = "../data/" + filename;
```
Comment (plibither8) [inline src/game-pregamemenu.cpp:142]: Let's be careful to remove commented out code

File: src/game.cpp [hunk 7]
```diff
@@ -28,13 +29,24 @@ enum GameStatusFlag {
   FLAG_ENDLESS_MODE,
   FLAG_GAME_IS_ASKING_QUESTION_MODE,
   FLAG_QUESTION_STAY_OR_QUIT,
+  FLAG_TILES_REMOVED,           // Indicates if tiles have already been removed
```
Comment (plibither8) [inline src/game.cpp:32]: Pedantic, but good for uniformity: let's remove the multiple spaces before the comment here. Best to move the comment above `FLAG_TILES_REMOVED`.

File: src/game.cpp [hunk 8]
```diff
@@ -178,6 +220,12 @@ bool_gamestatus_t check_input_other(char c, gamestatus_t gamestatus) {
       is_invalid_keycode = false;
     }
     break;
+  case CODE_RETURN_TO_MENU:
+    // When CODE_RETURN_TO_MENU is detected, the main menu is set to start, and game flags are adjusted.    mainmenustatus[FLAG_START_MENU] = true;
```
Comment (plibither8) [inline src/game.cpp:224]: I see commented out code 👀

File: src/game.cpp [hunk 9]
```diff
@@ -267,6 +315,21 @@ bool continue_playing_game(std::istream &in_os) {
...
+ * This function evaluates the current game status flags and takes appropriate actions:
+ * - It handles winning conditions and prompts the user to continue playing.
+ * - It checks for the end of the game conditions.
+ * - It manages saving the game state by prompting the user for a filename.
+ * - It resets the question asking event trigger for a new loop cycle.
+ * 
+ * @param gsgb A tuple containing the current game status and game board.
+ * @return bool_gamestatus_t A tuple containing a boolean indicating whether to continue the game loop and the updated game status flags.
+ * 
+ * @note Changes in the new version:
```
Comment (plibither8) [inline src/game.cpp:330]: Same with the note here! My suggestion: move these notes to the PR description.

File: src/gameboard.cpp [hunk 10]
```diff
@@ -135,7 +149,7 @@ bool canMoveOnGameboardDataArray(gameboard_data_array_t gbda) {
       const auto offset_check = {
           current_point + offset, // Positive adjacent check
           current_point - offset}; // Negative adjacent Check
-      for (const auto current_offset : offset_check) {
+      for (const auto &current_offset : offset_check) { //auto REFERENCE current_offset
```
Comment (plibither8) [inline src/gameboard.cpp:152]: commented out code 👀

File: src/loadresource.cpp [hunk 11]
```diff
@@ -14,27 +14,74 @@ namespace Game {
...
+      std::cerr << "Error. Cannot open the file: " << filename << std::endl;
+      return -1; // Return -1 to indicate an error
+  }
+
+  std::string tempLine;
+  int noOfLines = 0;
+
+  while (std::getline(stateFile, tempLine)) {
+      if (tempLine.find('[') != std::string::npos) {
+          break; // Exit loop if '[' is found
```
Comment (plibither8) [inline src/loadresource.cpp:40]: Great idea here, as well as the `-1` to return an error!

File: src/loadresource.cpp [hunk 12]
```diff
@@ -81,7 +128,7 @@ get_and_process_game_stats_string_data(std::istream &stats_file) {
   if (stats_file) {
     unsigned long long score;
     long long moveCount;
-    for (std::string tempLine; std::getline(stats_file, tempLine);) {
+    for (std::string tempLine; std::getline(stats_file, tempLine, '.');) {
```
Comment (plibither8) [inline src/loadresource.cpp:131]: Why was the `'.'` required?

File: src/menu.cpp [hunk 13]
```diff
@@ -146,9 +228,26 @@ bool soloLoop() {
...
+ */
 void endlessLoop() {
-  while (soloLoop())
-    ;
+  while (!mainmenustatus[FLAG_EXIT_GAME]) // As long as the exit option is not selected in the menu
+  {
+    mainmenustatus[FLAG_START_MENU] = true; // Initial state is Menu
+    if (mainmenustatus[FLAG_START_MENU] == true)    // If the menu flag is set, we enter...
+    {
+      soloLoop();                         // ... the soloLoop() function, where we navigate the menu
```
Comment (plibither8) [inline src/menu.cpp:248]: Let's move these comments to the line above.

File: src/scores-graphics.cpp [hunk 14]
```diff
@@ -56,7 +56,7 @@ std::string ScoreboardOverlay(scoreboard_display_data_list_t sbddl) {
              << "\n";
     };
 
-    for (const auto s : sbddl) {
+    for (const auto& s : sbddl) { // auto REFERENCE s
```
Comment (plibither8) [inline src/scores-graphics.cpp:59]: commented out code 👀

**Review by plibither8 — APPROVED** [review]
Code looks good, thank you very much! :)

I can see that a [CI workflow is failing](https://github.com/plibither8/2048.cpp/actions/runs/9585418900/job/26544757841?pr=145). Could you check the error behind it? Ideally, we want these checks to be successful before trusting it to be merged.

Additionally, a couple of changes I'd love to see before merging it:
* Let's remove the `.vscode` directory. More details in the review comments.
* Let's be uniform with how we do in-line comments. Gives much better readability.
* The added code would benefit from being run against a formater/linter. This will format the code to make it uniform. We've got a [`.clang-format`](https://github.com/plibither8/2048.cpp/blob/master/.clang-format) file that specifies the formatting conventions used in this project.

Nevertheless, I'm approving this PR on the basis of the solid feature code additions -- they look great!

**Comment by BurhanHTW** [comment]
We‘ll try to fix the CI. Hopefully this solved asap :)

**Comment by BurhanHTW** [comment]
I have to switch from C++14 to C++17 in CMakeLists.txt #include <filesystem> in menu.cpp is only supported from C++17 onwards.  https://en.cppreference.com/w/cpp/filesystem

**Comment by plibither8** [comment]
> I have to switch from C++14 to C++17 in CMakeLists.txt #include <filesystem> in menu.cpp is only supported from C++17 onwards.  https://en.cppreference.com/w/cpp/filesystem

Have you tried fixing the issue using the suggestion in the error message? `std::_fs::filesystem`

**Comment by BurhanHTW** [comment]
> > I have to switch from C++14 to C++17 in CMakeLists.txt #include  in menu.cpp is only supported from C++17 onwards.  https://en.cppreference.com/w/cpp/filesystem
> 
> Have you tried fixing the issue using the suggestion in the error message? `std::_fs::filesystem`

Not yet but what will it Change to have an _ ?

**Comment by plibither8** [comment]
@BurhanHTW could you push the changes to your branch to re-trigger the CI?

**Comment by BurhanHTW** [comment]
> @BurhanHTW could you push the changes to your branch to re-trigger the CI?

I will Push it later when i am at home. We have changed everything what you wanted :)

**Comment by BurhanHTW** [comment]
I cant push anything because the access is denied

**Comment by BurhanHTW** [comment]
I had to generate a new Token to push. Now we are waiting for the workflow approval. Hopefully its done :D

**Comment by plibither8** [comment]
@BurhanHTW your latest commit has many merge conflicts present. [Example](https://github.com/plibither8/2048.cpp/pull/145/commits/054d427eaa0541892a501adb8d9199c796e07f5d#diff-9177f1b4af50ab58a32411f440af9c4bb4f99beabe96721c913a0d8e9dac605dR127)

Remove all merge markers - they're causing the builds to fail.

**Comment by BurhanHTW** [comment]
> @BurhanHTW your latest commit has many merge conflicts present. [Example](https://github.com/plibither8/2048.cpp/pull/145/commits/054d427eaa0541892a501adb8d9199c796e07f5d#diff-9177f1b4af50ab58a32411f440af9c4bb4f99beabe96721c913a0d8e9dac605dR127)
> 
> Remove all merge markers - they're causing the builds to fail.

hopefully done :D

**Comment by BurhanHTW** [comment]
I don’t know what i have Done actually i have to Check the Problems …

**Comment by BurhanHTW** [comment]
As you can see in the commits i have deleted all markers. Maybe it will run without Problems now.

**Comment by BurhanHTW** [comment]
If i have to many Problems. Is it possible to make a new pull Request? I guess i had to many Problems because i commited to many files

**Comment by BurhanHTW** [comment]
If the workflow fails again i‘ll create a new Pull Request and Push the latest Code again. I am sorry to have those issues. :/

**Comment by BurhanHTW** [comment]
So the workflow failed again. I will create a new pull Request.

**Comment by plibither8** [comment]
Yep, that sounds like a good idea!

**Comment by BurhanHTW** [comment]
I have closed this pull request. I will create a new one now.

**Pull request closed without merging**